gnovm/pkg/doctest/analyzer.go

```diff
@@ -0,0 +1,205 @@
+	"strings"
+)
+
+// supported stdlib packages in gno.
+// ref: go-gno-compatibility.md
+var stdLibPackages = map[string]bool{
+	"bufio":           true,
+	"builtin":         true,
+	"bytes":           true,
+	"encoding":        true,
+	"encoding/base64": true,
+	"encoding/hex":    true,
+	"hash":            true,
+	"hash/adler32":    true,
+	"io":              true,
+	"math":            true,
+	"math/bits":       true,
+	"net/url":         true,
+	"path":            true,
+	"regexp":          true,
+	"regexp/syntax":   true,
+	"std":             true,
+	"strings":         true,
+	"time":            true,
+	"unicode":         true,
+	"unicode/utf16":   true,
+	"unicode/utf8":    true,
+
+	// partially supported packages
+	"crypto/cipher":   true,
+	"crypto/ed25519":  true,
+	"crypto/sha256":   true,
+	"encoding/binary": true,
+	"errors":          true,
+	"sort":            true,
+	"strconv":         true,
+	"testing":         true,
+}
+
+// analyzeAndModifyCode analyzes the given code block, adds package declaration if missing,
+// ensures a main function exists, and updates imports. It returns the modified code as a string.
+func analyzeAndModifyCode(code string) (string, error) {
+	fset := token.NewFileSet()
+	node, err := parser.ParseFile(fset, "", code, parser.AllErrors)
+	if err != nil {
+		// append package main to the code and try to parse again
+		node, err = parser.ParseFile(fset, "", "package main\n"+code, parser.ParseComments)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse code: %w", err)
+		}
+	}
+
+	ensurePackageDeclaration(node)
+	if exist := ensureMainFunction(node); !exist {
+		return "", fmt.Errorf("main function is missing")
+	}
+	updateImports(node)
+
+	src, err := codePrettier(fset, node)
+	if err != nil {
+		return "", err
+	}
+
+	return src, nil
+}
+
+// ensurePackageDeclaration ensures the code block has a package declaration.
+// by adding a package main declaration if missing.
+func ensurePackageDeclaration(node *ast.File) {
+	if node.Name == nil {
+		node.Name = ast.NewIdent("main")
+	}
+}
+
+// ensureMainFunction checks if a main function exists in the AST.
+// It returns an error if the main function is missing.
+func ensureMainFunction(node *ast.File) bool {
+	for _, decl := range node.Decls {
+		if fn, isFn := decl.(*ast.FuncDecl); isFn && fn.Name.Name == "main" {
+			return true
+		}
+	}
+	return false
+}
+
+// detectUsedPackages inspects the AST and returns a map of used stdlib packages.
+func detectUsedPackages(node *ast.File) map[string]bool {
+	usedPackages := make(map[string]bool)
+	remainingPackages := make(map[string]bool)
+	for pkg := range stdLibPackages {
+		remainingPackages[pkg] = true
+	}
+
+	ast.Inspect(node, func(n ast.Node) bool {
+		if len(remainingPackages) == 0 {
+			return false
+		}
+
+		selectorExpr, ok := n.(*ast.SelectorExpr)
+		if !ok {
+			return true
+		}
+
+		ident, ok := selectorExpr.X.(*ast.Ident)
+		if !ok {
+			return true
+		}
+
+		if remainingPackages[ident.Name] {
```

Comment:
Very tricky one. I think the approach could be to fist traverse the AST to collect local identifiers (functions, variables...), and then traverse AST again to examine each node, treating an identifier as the stdlib only if it's not in the list of local identifiers. 🤔🤔🤔🤔

Reply:
it works https://github.com/gnolang/gno/pull/2357/commits/317fd41c20cf0837eb618d382cd64bd1a80bf7ce

Reply:
I think that, in order to know whether or not to add a stdlib import, we need to know the local identifiers that are defined in the current scope. The current scope would be any package level identifiers, block level identifiers, and ancestor block level identifiers (I think?).  For example, this code will import the `math` stdlib when it should not:
```go
package main
			
type mmm struct{
	MaxUint64 uint64
}

func main() {
	maxValue := getMaxValue()
	println(maxValue)
}

func getMaxValue() uint64 {
	math := mmm{MaxUint64: 1<<64 - 1}
	return math.MaxUint64
}
```

Reply:
I wonder if these feature should be omitted for the time being, making it required that people include the necessary imports when writing code. This is a nice to have but I think it could be difficult to do correctly.

Reply:
It would be nice if it could be handled well, but there certainly seems to be some risk. So, as mentioned below, I think it's appropriate to exclude this part. So will removed.